### ISSUE TITLE
[bug] wrong processing of hostnames with UDP ASSOCIATE

### ISSUE THREAD

**Issue by ezbik** [3proxy/3proxy#952]
Конфиг

```
nserver 127.0.0.53
socks -a -n -p1080  -l
```

проверяем UDP скриптом на Питоне https://termbin.com/gvz38 . Кратко что делает скрипт - запрашивает UDP ASSOCIATE на **Хостнейм** dns.google:53 и потом туда шлет DNS запрос (IN A baidu.com)  и читает ответ.

Ожидаемый результат - 3proxy разрезолвит запрашиваемое имя dns.google через свой nserver и откроет UDPMAP.

Реальный результат - 3proxy шлет неверный DNS запрос на nserver, пытается разрезолвить странное имя "0" (ноль), словил tcpdump'ом:

```
14:17:00.931676 IP 127.0.0.1.47672 > 127.0.0.53.53: 18691+ A? 0. (19)
14:17:00.931804 IP 127.0.0.53.53 > 127.0.0.1.47672: 18691 ServFail* 0/0/0 (19)
```

Костыль/фикс:

- убрать nserver, тогда 3proxy нормально резволит имя через **тот же**  dns, но беря его из  /etc/resolv.conf .

**Comment by z3APA3A** [comment]
Скорей всего скрипт вместо IP адреса 0.0.0.0 шлет хостнейм как строку "0.0.0.0", посмотрю почему это не работает на стороне 3proxy, но в целом в скрипте здесь тоже очевидный баг

**Comment by z3APA3A** [comment]
Т.е. запрос к nserver с "0" связан не с разрешением того запроса что прилетел поверх SOCKS, а с разрешением имени которое прилетело в SOCKS-команде UDPASSOC. Скорей всего встроенный резолвер такой запрос нормально обрабатывает и резолвит в 0.0.0.0, а сервер который прописан в nserver дает SERVFAIL.

**Comment by ezbik** [comment]
Я вижу что на порт tcp/1080 прилетает "0".

```
Socks Protocol
    Version: 5
    Command: UdpAssociate (3)
    Reserved: 0
    Address Type: Domain Name (3)
    Remote name: 0
    Port: 38189
```

Далее по UDP прилетает уже имя.

```
Socks
    Reserved: 0
    Fragment Number: 0
    Address Type: Domain Name (3)
    Remote name: dns.google
    Remote Port: 53
```

**Comment by ezbik** [comment]
да, я не смог найти в RFC 1928 , что имя может быть "0" в момент UdpAssociate. Но эта ситуация обрабатывается  хорошо такими прокси как gost.run, glider, v2ray.

**Comment by FastSpace123** [comment]
> но в целом в скрипте здесь тоже очевидный баг: поддерживать IP адрес переданный как hostname SOCKS сервер строго говоря не обязан, в hostname ожидается именно DNS-имя

Есть сайт https://networktest.twilio.com/
Чтобы пройти тест UDP клиент шлёт hostname как IP адрес

**Comment by FastSpace123** [comment]
> @FastSpace123 3proxy поддерживает передачу dotted-IP в хостнейме, здесь клиент вообще передает адрес как строку в integer-форме ("0").

Ясно, но у меня ситуация такая же как и у ezbik. UDP проверку на сайте проходит браузер толькоесли вырублю из конфига nsserver настройку и оставлю resolv.conf.

**Comment by FastSpace123** [comment]
Вот, что написал разработчик софта, про тест выше. Я беспонятия, что там написано. Может до вас дойдет

Тест заключается в том, что передача данных осуществляется через TURN сервер с на локальный пк. То есть звонок самому себе.
Проблема проявляется только если использовать домен в url TURN сервера.
Если использовать ip, то все будет работать корректно.
Для того чтобы найти TurnPort(это название класса из исходника браузера), браузер использует ip предварительно полученный из DNS запроса.
Библиотека для туннелирования заменяет ip на 10.12.0.0(или что-то подобное) и откладывает DNS запрос для того, чтобы отправить его позднее в прокси сервер.
Причины проблемы в том, что правильный TurnPort не может быть найден, ведь 10.12.0.0 сравнивается с реальным ip.
Чтобы исправить, алгоритм для поиска TurnPort был заменен. Теперь transaction id из последнего запроса Allocate Request сверяется с transaction id из прешедшего пакета.
https://i.imgur.com/cm3in0s.png
Как только порт найден, временный порт заменяется на настоящий, и дальше алгоритм работает как раньше.
Еще один момент, далеко не все прокси умеют работать с UDP.
И даже среди тех которые умеют, не все поддерживают DNS запросы на стороне прокси сервера.
Поэтому фикс будет работать не везде."

**Comment by z3APA3A** [comment]
Это все не имеет отношения к запросам к socks.

**Comment by FastSpace123** [comment]
> Это все не имеет отношения к запросам к socks.

Ясно, спасибо за ответ. 

Стоит перейти на другой значит клиент значи, мне нужно совершить звонок. 3proxy это не умеет.

**Comment by z3APA3A** [comment]
Попробуй текущий мастер с dc7e098e6b5563e1f31bdd1ad165a64f4983fd3c, поломанные клиенты должны нормально работать

**Comment by ezbik** [comment]
@z3APA3A  Спасибо, заработало!

Но я вижу, что не работают ACL , пример 

```
nserver 1.1.1.1
auth strong
deny * * 8.8.0.0/16
allow *
socks -a -n -p1080  -l
users alice:CL:cool
```

пропускает по socks/UDP как dns.google так и 8.8.8.8.

**Comment by z3APA3A** [comment]
Да, по UDP клиент одновременно может работать со всеми серверами через одно подключение к SOCKS. Сейчас авторизация проверяется один раз, если в назначении рзрешен адрес 0.0.0.0 или прописанный extip, то будет разрешено отправлять UDP пакеты на любые IP адреса, потому что проверять авторизацию для каждого UDP пакета это очень дорого. По той же причине в логах будет только одна запись и только один внешний IP (последний)
